rfc/rfc-91/rfc-91.md

```diff
@@ -0,0 +1,145 @@
+- https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html
+
+When we create the new lock file in tryLock we will use the If-None-Match precondition. From AWS docs:
+- *Uploads the object only if the object key name does not already exist in the bucket specified. Otherwise, Amazon S3 returns a 412 Precondition Failed error. If a conflicting operation occurs during the upload S3 returns a 409 ConditionalRequestConflict response. On a 409 failure you should retry the upload. Expects the '\*' (asterisk) character.*
+
+#### Etags
+
+- https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html
+
+Etags are unique hashes of the contents of the object. Since our payload has a unique owner uuid, as long as the expiration (which is calculated by System.currentTimeMillis()) changes across requests for the same node, the Etag will change (otherwise the request would return 304 instead of 201/202).
+
+When we overwrite an existing file in any of the methods, we will use the If-Match precondition. From AWS docs:
+- *Uploads the object only if the ETag (entity tag) value provided during the WRITE operation matches the ETag of the object in S3. If the ETag values do not match, the operation returns a 412 Precondition Failed error. If a conflicting operation occurs during the upload S3 returns a 409 ConditionalRequestConflict response. On a 409 failure you should fetch the object's ETag and retry the upload. Expects the ETag value as a string.*
+
+#### GCP/GCS
+
+- https://cloud.google.com/storage/docs/request-preconditions
+- https://cloud.google.com/storage/docs/metadata#generation-number
+
+GCS has ETags, but they also have generation numbers, which are even better, and work for more use cases. Our current implementation already uses them, so they do not need further validation.
+
+When we create the new lock file in tryLock we will use generationMatch(0). From GCP docs:
+- *Passing the if_generation_match parameter to a method which retrieves a blob resource (e.g., Blob.reload) or modifies the blob (e.g., Blob.update) makes the operation conditional on whether the blob’s current generation matches the given value. As a special case, passing 0 as the value for if_generation_match makes the operation succeed only if there are no live versions of the blob.*
+
+We can use the same logic for preconditions with overwrite operations using the currently stored lock file's generation number.
+
+## Rollout/Adoption Plan
+
+ - What impact (if any) will there be on existing users?
+   - None 
+ - If we are changing behavior how will we phase out the older behavior?
+   - N/A
+ - If we need special migration tools, describe them here.
+   - N/A
+ - When will we remove the existing behavior
+   - N/A
+
+## Test Plan
+
+We can write normal junit tests using testcontainers with GCS and S3 to simulate edge cases and general contention. Further adhoc testing will include the following scenarios:
```

Comment:
Should we add a stress test purely on the new lock provider, i.e., creating a large number of concurrent threads, e.g., 1000, with each trying to lock and unblock with a random wait time of 0-1 second in between and validating that only one thread should hold the lock at any given time to make sure the implementation works as expected under extreme parallelism?

Reply:
yes this will be added, and updating test plan to reflect